cranelift/codegen/src/opts/algebraic.isle

```diff
@@ -537,3 +537,25 @@
 ;; both of them since they cancel out.
 (rule (simplify (fmul ty (fneg ty x) (fneg ty y)))
       (fmul ty x y))
+
+
+;; Wasmtime generates a lot of load+bitcast's. This is because in Wasm
+;; SIMD loads are always v128.load, without type information. This means
+;; that the translation to CLIF first loads into a `i8x16` and then each
+;; op bitcasts the vector as needed.
+;;
+;; This bitcasting prevents the x64 backend from merging the load into the
+;; op, so try to change the type of the load into the bitcast type.
+(rule (simplify (bitcast cty cflags (load lty lflags addr offset)))
+      ;; Two preconditions for this rule:
+      ;; 1. The types must have the same size. (this is enforced by the verifier)
+      ;; 2. The endiannes on both memflags must match
+      (if-let $true (mem_flags_endianness_eq cflags lflags))
+      (load cty lflags addr offset))
+
+;; TODO: We could also do this for `store`. It doesen't seem to be generated in clif_opt.isle.
```

Comment:
We can't currently write egraph rules that will replace an instruction unless it has exactly one result `Value`, which is why `store` doesn't work here. I recently removed the other instructions from `clif_opt.isle` because rules using them would never fire. We've been talking about how to do this but don't have a concrete plan yet.

This restriction is primarily because our egraph identifies equivalence classes of `Value`s, rather than identifying equivalent instructions. It's also because it's harder to define what equivalence means for instructions with side effects, such as `store`.